README

                  Gnatsweb - web interface to GNATS

                            Dirk Bergstrom
                             Kenneth Cox
                           Matt Gerassimoff
                            Yngve Svendsen


What is Gnatsweb?
=================

    Gnatsweb is a web interface to GNATS, the GNU Problem Report
    Management System.  It is a Perl CGI script which runs on your web
    server.


System Requirements
===================

    Gnatsweb 4 requires that you have the following:

    * The 'gnatsd' network daemon installed and configured, version
      4.0 beta 2 or greater.

      IMPORTANT: Gnatsweb 4.x will NOT work with GNATS 3.x

    * Write access to a CGI-enabled directory on your web server.

    * Perl 5, with the fine CGI.pm module, version 2.56 or newer.
      Note that the unofficial 3.x versions of CGI.pm have been
      reported not to work. There has been trouble reported with a
      version of perl prior to 5.004_04.


Where do I get it?
==================

    ftp://ftp.gnu.org/pub/gnu/gnatsweb/

    It's also available in the contrib/gnatsweb directory of the GNATS
    distribution, but since GNATS and Gnatsweb releases aren't
    necessarily synchronized, this version will often be outdated.

    Note that versions of Gnatsweb numbered 4.x only work with 4.x
    versions of GNATS.

How do I install it?
====================

    Read the INSTALL file.  For postinstall customization, read
    CUSTOMIZE, then CUSTOMIZE.vars and CUSTOMIZE.cb.


Who wrote it?
=============

    Gnatsweb was originally written by Matt Gerassimoff and Kenneth
    H. Cox.  It is currently being maintained by Yngve Svendsen.  Dirk
    Bergstrom made very significant contributions.

    Grateful acknowledgement is given to the authors of wwwgnats, a
    precursor to gnatsweb in spirit if not in code base: Dan Kegel and
    Huy Le, Brian Behlendorf and Ken Coar.


Licensing
=========

    GNU Gnatsweb is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2, or
    (at your option) any later version.

    GNU Gnatsweb is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Gnatsweb; see the file COPYING. If not, write to the
    Free Software Foundation, Inc., 59 Temple Place - Suite 330,
    Boston, MA 02111-1307, USA.

    Copyright 1998, 1999, 2001, 2003
    - The Free Software Foundation Inc.


NEWS
Current Version: 2.9.3
See the file ChangeLog for the details of all changes.

Changes from 2.9.2 to 2.9.3
* We now display indented material in multiline fields, such as code
  samples etc., with the indentation intact.
* We now include a Reply-To header in change notification emails, set
  up in such a way that followups to change notification e-mails are also
  sent to GNATS itself.
* After successful PR submission, the user is now redirected to the
  page the user was at when "Create" was clicked.
* Support queries on Submitter-ID from the standard query page.
* The redirect after stored query deletion now works in Netscape.
* The "send mail to interested parties" link now includes the PR
  synopsis as well as category and PR number in the Subject field of
  the generated message.

Changes from 2.9.1 to 2.9.2
* Workaround for an age-old stupid, stupid, stupid Netscape bug, in
  which the back arrow is broken after submitting a new PR or an
  edited PR.
* Use the $include_audit_trail variable to choose view command for the
  PR number links in query results.  This was mistakenly left out in
  2.9.1.
* Documentation polished.

Changes from 2.9.0 to 2.9.1
* Documentation corrected and significantly expanded.  We now provide
  a TROUBLESHOOTING document as well.
* Bug fix. E-mail address extraction for the "send email to interested
  parties" is now more robust.
* Bug fix. Handle browsers without Javascript support better.
* Bug fix. Minor problems with browsers such as Konqueror should be
  gone.
* Minor new feature: We now use both the category name and description
  and the responsible name and full name in dialogs on the create,
  edit and query pages.
* Minor new feature: Whether the Audit Trail is visible by default in
  the View PR screen is now configurable.
* Minor new feature: There is now a checkbox in the query pages which
  provides for displaying the current date on the query results page.

Changes from 2.8.2 to 2.9.0
* Usability enhancement. There is now a navigation bar on top of all
  pages with links to all major Gnatsweb functions.
* Bug fix. Properly handle a locked GNATS database. This used to make
  Gnatsweb bomb out when submitting edits and leave the PR locked.
* Minor bug fixes related to attachments.
* Add X-Mailer header to outgoing mail.
* Bug fix. Text attachments are no longer truncated at the first blank
  line.

Changes from 2.8.1 to 2.8.2
* Fixed serious security hole allowing a attacker to read files and
  execute commands remotely on the gnatsweb server. See ChangeLog for
  the full monty.

Changes from 2.8.0 to 2.8.1
* Redirect after submitting edits wasn't working on systems with a
  new CGI.pm. This has been fixed.
* We now let people search for PRs with category 'pending'.
* We no longer by default require JavaScript for the main menu to
  work.
* The INSTALL file now makes it abundantly clear that you need a
  gnats-db.conf file even if you have only a single database.

Changes from 2.7 beta to 2.8.0
* Both the Gnats and Gnatsweb version numbers are now displayed in the
  main page footer.
* If a user pressed the 'Stop' button of the browser right after
  pressing 'Submit' on the Edit PR page, the PR would not be properly
  unlocked. This has been fixed in a hopefully watertight way.
* If edits were submitted to a locked PR, Gnatsweb would not catch the
  error, and instead display a confusing error message saying that the
  PR had been modified since you started editing it. Gnatsweb now
  properly warns of a locked PR and stops the submission.
* Documentation updates to avert problems when people install Gnatsweb
  on top of a multiple-database Gnats.
* The page banner is now a little smaller, and the text "gnatsweb" is
  now displayed in a smaller typeface.
* On the Advanced Query Page, we now display a line below 'Closed
  After' and 'Closed Before' reminding users to uncheck 'Ignore
  Closed'.
* A bug where users would be able to create new PRs with PR numbers
  composed both out of letters and numbers has been corrected.
* When a non-existing PR number was typed into the 'view' or 'edit'
  textboxes on the front page, no error message would be given, and
  the user would be presented with a blank PR. Checks have been added
  to prevent this.
* Internet Explorer users are now being properly redirected to the
  Gnatsweb main page after submitting an edit.
* Internet Explorer would try to give downloaded attachments the same
  name as the gnatsweb script file itself. This has been fixed.
* Editing PRs with attachments caused Gnatsweb 2.7 beta to crash. This
  has been fixed.
* URLs in PR text are now being automatically converted to clickable
  links when PRs are displayed.
* Fixed a bug where HTML tags in PR fields weren't being properly
  escaped before being displayed in the query results.
* It is now possible to link an external stylesheet to all Gnatsweb
  pages. The Gnatsweb-generated HTML has been thoroughly cleaned up in
  order to support stylesheets better.


gnatsweb-site-example.pl
# gnatsweb-site.pl -
#     Gnatsweb site-specific variables and subroutines.
#
# Copyright 1998, 1999, 2001, 2003
# - The Free Software Foundation Inc.
#
# GNU Gnatsweb is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# GNU Gnatsweb is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Gnatsweb; see the file COPYING. If not, write to the Free
# Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: gnatsweb-site-example.pl,v 1.9.2.1 2003/07/29 12:24:22 yngves Exp $
#

# GNATS host.
$site_gnats_host = 'localhost';

# The port which the GNATS server is running on.
# Default installations of GNATS run on port 1529.
$site_gnats_port = '1529';

# Subroutine for customizing setup and behaviour.
# Gnatsweb customizations should be done here instead of in the main script.
# This isolates changes and makes for easier upgrades.
# We provide a range of example callbacks below.
sub site_callback {
    my($reason, @args) = @_;

    # Use warn() to facilitate debugging.
    # Look for the messages in the web server error log.
    #warn "site_callback: @_\n";


	# The following callback routine creates a form button which takes
	# the user to the Create PR page with Class set to 'support' and
	# Submitter-Id set to 'external'.  It is called from inside the
	# main_page subroutine of gnatsweb.pl.
    if ($reason eq 'main_page_top_buttons')
	{
		my $html = one_line_form('Open Support Issue:',
					  $q->submit(-name=>'cmd', -value=>'create'),
					  $q->hidden(-name=>'Class', -default=>'support'),
					  $q->hidden(-name=>'Submitter-Id', -default=>'internal'));
		return $html;
	}

    # The following routine is similar to the previous one.  It
    # creates two buttons for the bottom of the main page.  The first
    # button allows the user to search directly for all open PRs, the
    # second provides a direct search for all non-closed PRs.  Note
    # that the buttons submit the commands 'open' and 'not closed'
    # respectively.  These are commands that aren't supported in the
    # default gnatsweb.pl, but below, we use the 'cmd' callback hook
    # to provide these commands.
    elsif ($reason eq 'main_page_bottom_buttons')
    {
        my $html = one_line_form('Direct search:',
                            $q->submit(-name=>'cmd', -value=>'open'),
                            '&nbsp;',
						    $q->submit(-name=>'cmd', -value=>'not closed'));
        return $html;
    }

	# This routine is called from the page_start_html subroutine of
	# gnatsweb.pl.  It overrides the default HTML written on top of
	# each page, i.e. the banner and the button bar, replacing it with
	# a very simplistic banner.
	elsif ($reason eq 'page_start_html')
	{
		my $title = $args[0];
		my $html = $q->b(uc("$title - $site_banner_text")) . $q->hr;
		return $html;
	}

    # Construct the HTML which will be printed just below the heading
    # on the login page.
    elsif ($reason eq 'login_page_text')
    {
        my $html = $q->p('This is the GNATS bug tracking system used for
                          reporting bugs in our defrobulator products.') .
                   $q->p("Use your password and username to log in.
                          If you haven't been assigned a login,
                          please contact " .
                   $q->a({-href=>'mailto:helpdesk@example.com'},
                          'helpdesk@example.com'));
        return $html;
    }

	# Construct HTML which will be added above the Description field
    # in the View and Edit pages.
    elsif ($reason =~ /^(sendpr|edit)_intro_Description$/)
    {
        my $stuff = "<b>Description should include contact info<b>"
            . "<br>";
        return $stuff;
    }

	# The following block of code is called from the end of the main
	# routine of gnatsweb.pl.  It allows us to define custom commands,
	# in this case 'open' and 'not closed'.  The first searches for
	# all open PRs and the second searches for all non-closed PRs.  We
	# use this commands to provide direct searches from the main page,
	# see above.
    elsif ($reason eq 'cmd')
    {
        my $cmd = $args[0];

        if ($cmd eq 'open')
        {
            # Direct search for open PRs
			print_header(-cookie => create_global_cookie());
            initialize();
            my $page = 'Query Results';
            my $heading = 'Query Results';
            page_start_html($page);
            page_heading($page, $heading);
            $q->param(-name=>'columns',-values=>['Category','Synopsis','Responsible',
                                                 'State','Arrival-Date'
                                                 ]);
            client_cmd("rset");
            client_cmd('qfmt "%s^_%d^_%s^_%d^_%d^_%{%Y-%m-%d %H:%M:%S %Z}D" ' . 
                       'builtinfield:Number Category Synopsis Responsible State Arrival-Date');
			client_cmd("expr State~\"open\"");
            my(@query_results) = client_cmd("quer");
            display_query_results(@query_results);
            page_footer($page);
            page_end_html($page);
            exit;
        }

		elsif ($cmd eq 'not closed')
		{
            # Direct search for non-closed PRs
			print_header(-cookie => create_global_cookie());
			initialize();
			my $page = 'Query Results';
			my $heading = 'Query Results';
			page_start_html($page);
			page_heading($page, $heading);
            $q->param(-name=>'columns',-values=>['Category','Synopsis','Responsible',
                                                 'State','Arrival-Date'
                                                 ]);
            client_cmd("rset");
            client_cmd('qfmt "%s^_%d^_%s^_%d^_%d^_%{%Y-%m-%d %H:%M:%S %Z}D" ' . 
                       'builtinfield:Number Category Synopsis Responsible State Arrival-Date');
			client_cmd("expr ((! builtinfield:State[type]=\"closed\"))");
			my(@query_results) = client_cmd("quer");
			display_query_results(@query_results);
			page_footer($page);
			page_end_html($page);
			exit;
		}
	}
}
